astropy/units/core.py

```diff
@@ -2262,13 +2262,34 @@ class CompositeUnit(UnitBase):
 
     _decomposed_cache: CompositeUnit | None = None
 
+    @overload
+    def __init__(
+        self,
+        scale: Complex,
+        bases: Sequence[UnitBase],
+        powers: Sequence[Real],
+        decompose: bool = False,
+        decompose_bases: Collection[UnitBase] = (),
+        _error_check: Literal[True] = True,
+    ) -> None: ...
+    @overload
+    def __init__(
+        self,
+        scale: UnitScale,
+        bases: Sequence[UnitBase],
```

Comment:
We could make this `list[UnitBase]` and `list[UnitPower]`, since that is what is used by units internally, which is all that should get here.

Reply:
You should read [PEP 483 – The Theory of Type Hints](https://peps.python.org/pep-0483/) and in particular the section about [Covariance and Contravariance](https://peps.python.org/pep-0483/#covariance-and-contravariance). The relevant takeaway for the annotations in question is that `Sequence` is covariant, but `list` is invariant, so `Sequence` is correct and `list` is not.